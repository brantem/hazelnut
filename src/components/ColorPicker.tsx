import { RadioGroup } from '@headlessui/react';
import clsx from 'clsx';

import colors from 'data/colors';

type ColorPickerProps = {
  value: string;
  onChange: (color: string) => void;
  isDisabled?: boolean;
};

const ColorPicker = ({ value, onChange, isDisabled }: ColorPickerProps) => {
  return (
    <RadioGroup value={value} onChange={onChange} data-testid="color-picker">
      <RadioGroup.Label className="block text-sm text-neutral-700">Color</RadioGroup.Label>

      <div className="mt-2 flex w-full justify-between space-x-2">
        {colors.map((color) => {
          const isActive = color === value;
          return (
            <RadioGroup.Option
              value={color}
              key={color}
              className={clsx(
                `h-8 w-full rounded-full ring-offset-2 focus:outline-none md:h-10`,
                isActive && 'ring-2',
                isDisabled ? `bg-${color}-300 ring-neutral-300` : `bg-${color}-500 ring-${color}-500 cursor-pointer`,
              )}
              disabled={isDisabled}
              data-testid={`color-picker-option-${color}`}
            />
          );
        })}
      </div>
    </RadioGroup>
  );
};

export default ColorPicker;
