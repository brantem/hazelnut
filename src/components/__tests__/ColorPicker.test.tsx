import { render, screen, act } from '@testing-library/react';
import '@testing-library/jest-dom';

import ColorPicker from 'components/ColorPicker';

describe('ColorPicker', () => {
  it('should render successfully', () => {
    const onChange = vi.fn(() => {});
    const { container, rerender } = render(<ColorPicker value="" onChange={onChange} />);

    expect(container).toMatchSnapshot();
    expect(container.querySelector('[aria-checked="true"]')).not.toBeInTheDocument();
    const option = screen.getByTestId('color-picker-option-red');
    act(() => option.click());
    expect(onChange).toHaveBeenCalledWith('red');
    rerender(<ColorPicker value="red" onChange={() => {}} />);
    expect(option).toHaveAttribute('aria-checked', 'true');
    expect(container).toMatchSnapshot();
  });

  it("shouldn't be clickable if disabled", () => {
    const onChange = vi.fn(() => {});
    const { container } = render(<ColorPicker value="" onChange={onChange} isDisabled />);

    expect(container).toMatchSnapshot();
    screen.getByTestId('color-picker-option-red').click();
    expect(onChange).not.toHaveBeenCalled();
  });
});
