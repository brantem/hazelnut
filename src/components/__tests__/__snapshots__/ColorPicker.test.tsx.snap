// Vitest Snapshot v1

exports[`ColorPicker > should render successfully 1`] = `
<div>
  <div
    aria-labelledby="headlessui-label-:r1:"
    data-testid="color-picker"
    id="headlessui-radiogroup-:r0:"
    role="radiogroup"
  >
    <label
      class="block text-sm text-neutral-700"
      id="headlessui-label-:r1:"
      role="none"
    >
      Color
    </label>
    <div
      class="flex flex-wrap justify-between space-x-3 mt-2 w-full"
      role="none"
    >
      <div
        aria-checked="false"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-red-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state=""
        data-testid="color-picker-option-red"
        id="headlessui-radiogroup-option-:r2:"
        role="radio"
        tabindex="0"
      />
      <div
        aria-checked="false"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-amber-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state=""
        data-testid="color-picker-option-amber"
        id="headlessui-radiogroup-option-:r3:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-lime-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state=""
        data-testid="color-picker-option-lime"
        id="headlessui-radiogroup-option-:r4:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-emerald-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state=""
        data-testid="color-picker-option-emerald"
        id="headlessui-radiogroup-option-:r5:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-cyan-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state=""
        data-testid="color-picker-option-cyan"
        id="headlessui-radiogroup-option-:r6:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-blue-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state=""
        data-testid="color-picker-option-blue"
        id="headlessui-radiogroup-option-:r7:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-violet-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state=""
        data-testid="color-picker-option-violet"
        id="headlessui-radiogroup-option-:r8:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-fuchsia-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state=""
        data-testid="color-picker-option-fuchsia"
        id="headlessui-radiogroup-option-:r9:"
        role="radio"
        tabindex="-1"
      />
    </div>
  </div>
</div>
`;

exports[`ColorPicker > should render successfully 2`] = `
<div>
  <div
    aria-labelledby="headlessui-label-:r1:"
    data-testid="color-picker"
    id="headlessui-radiogroup-:r0:"
    role="radiogroup"
  >
    <label
      class="block text-sm text-neutral-700"
      id="headlessui-label-:r1:"
      role="none"
    >
      Color
    </label>
    <div
      class="flex flex-wrap justify-between space-x-3 mt-2 w-full"
      role="none"
    >
      <div
        aria-checked="true"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 ring-2 bg-red-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state="checked active"
        data-testid="color-picker-option-red"
        id="headlessui-radiogroup-option-:r2:"
        role="radio"
        tabindex="0"
      />
      <div
        aria-checked="false"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-amber-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state=""
        data-testid="color-picker-option-amber"
        id="headlessui-radiogroup-option-:r3:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-lime-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state=""
        data-testid="color-picker-option-lime"
        id="headlessui-radiogroup-option-:r4:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-emerald-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state=""
        data-testid="color-picker-option-emerald"
        id="headlessui-radiogroup-option-:r5:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-cyan-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state=""
        data-testid="color-picker-option-cyan"
        id="headlessui-radiogroup-option-:r6:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-blue-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state=""
        data-testid="color-picker-option-blue"
        id="headlessui-radiogroup-option-:r7:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-violet-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state=""
        data-testid="color-picker-option-violet"
        id="headlessui-radiogroup-option-:r8:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-fuchsia-500 ring-neutral-500 cursor-pointer"
        data-headlessui-state=""
        data-testid="color-picker-option-fuchsia"
        id="headlessui-radiogroup-option-:r9:"
        role="radio"
        tabindex="-1"
      />
    </div>
  </div>
</div>
`;

exports[`ColorPicker > shouldn't be clickable if disabled 1`] = `
<div>
  <div
    aria-labelledby="headlessui-label-:rb:"
    data-testid="color-picker"
    id="headlessui-radiogroup-:ra:"
    role="radiogroup"
  >
    <label
      class="block text-sm text-neutral-700"
      id="headlessui-label-:rb:"
      role="none"
    >
      Color
    </label>
    <div
      class="flex flex-wrap justify-between space-x-3 mt-2 w-full"
      role="none"
    >
      <div
        aria-checked="false"
        aria-disabled="true"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-red-300 ring-neutral-300"
        data-headlessui-state="disabled"
        data-testid="color-picker-option-red"
        id="headlessui-radiogroup-option-:rc:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        aria-disabled="true"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-amber-300 ring-neutral-300"
        data-headlessui-state="disabled"
        data-testid="color-picker-option-amber"
        id="headlessui-radiogroup-option-:rd:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        aria-disabled="true"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-lime-300 ring-neutral-300"
        data-headlessui-state="disabled"
        data-testid="color-picker-option-lime"
        id="headlessui-radiogroup-option-:re:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        aria-disabled="true"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-emerald-300 ring-neutral-300"
        data-headlessui-state="disabled"
        data-testid="color-picker-option-emerald"
        id="headlessui-radiogroup-option-:rf:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        aria-disabled="true"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-cyan-300 ring-neutral-300"
        data-headlessui-state="disabled"
        data-testid="color-picker-option-cyan"
        id="headlessui-radiogroup-option-:rg:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        aria-disabled="true"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-blue-300 ring-neutral-300"
        data-headlessui-state="disabled"
        data-testid="color-picker-option-blue"
        id="headlessui-radiogroup-option-:rh:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        aria-disabled="true"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-violet-300 ring-neutral-300"
        data-headlessui-state="disabled"
        data-testid="color-picker-option-violet"
        id="headlessui-radiogroup-option-:ri:"
        role="radio"
        tabindex="-1"
      />
      <div
        aria-checked="false"
        aria-disabled="true"
        class="h-10 w-10 rounded-full focus:outline-none ring-offset-2 bg-fuchsia-300 ring-neutral-300"
        data-headlessui-state="disabled"
        data-testid="color-picker-option-fuchsia"
        id="headlessui-radiogroup-option-:rj:"
        role="radio"
        tabindex="-1"
      />
    </div>
  </div>
</div>
`;
